### PR TITLE
feat: update run0edit to v0.4.3 w/ support for immutable flag

### PR DESCRIPTION
**Pull request by HastD**
The new version of [`run0edit`](https://github.com/HastD/run0edit) checks for the immutable flag on the file to be edited (or the directory if the file doesn't exist), and if so, asks the user whether to temporarily remove it to edit/create the file. This makes editing immutable files less error-prone, as the immutable flag is only removed for the brief window necessary to overwrite the file contents and reapplied immediately afterward; the script also compares the file contents after the immutable flag has been reapplied to ensure it was not modified by another program during that window.

There are also miscellaneous improvements to the script's argument handling, which now parses arguments in a more conventional way and accepts both `--help` and `--version` arguments, and the inner privileged script has been refactored to reduce code duplication.

### CODE REVIEW

File: files/system/usr/bin/run0edit [hunk 1]
```diff
@@ -71,21 +115,43 @@ else
 fi
 
 # Create a temporary file with a random suffix appended to the filename.
-filename="$1"
-directory=$(dirname "$filename")
-trunc_filename=$(basename "$filename" | head -c 64)
-temp_filename=$(mktemp --tmpdir "${trunc_filename}.XXXXXXXXXX")
+directory=$(dirname -- "$filename")
```
Comment (ShadowSlayer1441) [inline files/system/usr/bin/run0edit:122]: Should probably check if the file's directory is immutable, before trying to create temp file.

File: files/system/usr/bin/run0edit [hunk 2]
```diff
@@ -71,21 +115,43 @@ else
...
 script='\
+    ( set -o pipefail 2> /dev/null ) && set -o pipefail
+    set -eu
+    PATH=$(command -p env -i sh -c "echo \"\$PATH\"")
     filename="$1"
     tmpfile="$2"
     editor="$3"
+    edit_immutable=no
+
+    ask_immutable() {
```
Comment (ShadowSlayer1441) [inline files/system/usr/bin/run0edit]: Before asking to proceed with an immutable file flagged file, it should check if it's fs is mounted read only. (Otherwise user could assent to proceed only to get a error later anyway.) Something like this could work, theoretically I think you need to check both directory and file thanks to sim-links, but that's quite the edge case.
```elif [ -n "$(findmnt -no OPTIONS --target "$directory" | grep ro)" ]; then```

Also note none of these tests are privileged, so ideally they should be run as user before polkit authentication to save user time.

Comment (HastD) [reply files/system/usr/bin/run0edit]: Any test that requires read access to the file has to be run in the privileged script, as the file (or even the directory it's in) might not be readable by the user. Anyway, good idea to add the check for a read-only filesystem, I'll put that in.

File: files/system/usr/bin/run0edit (hunk 1 above)
Comment (HastD) [reply files/system/usr/bin/run0edit:122]: Empty temp files are cleaned up at the end of the script regardless, so I don't think this makes a difference.

Comment (ShadowSlayer1441) [reply files/system/usr/bin/run0edit:122]: It will error out if the directory is immutable because it will fail to create a file, right?

Comment (HastD) [reply files/system/usr/bin/run0edit:122]: Oh I see what you're saying. Yes, I'll add support for handling the immutable flag on the directory in the case where the target file doesn't exist yet. However, this doesn't affect the temp file because it's created in `/tmp`, not in the target directory.

File: files/system/usr/bin/run0edit [hunk 3]
```diff
@@ -30,29 +31,72 @@ set -eu
...
+    case "$1" in
+        -v|--version)
+            echo "run0edit $version"
+            exit
+            ;;
+        -h|--help)
+            print_usage
+            exit
+            ;;
+        --)
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit:58]: will this ever be hit since getopts isn't being used?

File: files/system/usr/bin/run0edit [hunk 4]
```diff
@@ -65,33 +109,89 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
 script='\
+    ( set -o pipefail 2> /dev/null ) && set -o pipefail
+    set -eu
+    PATH=$(command -p env -i sh -c "echo \"\$PATH\"")
     filename="$1"
     tmpfile="$2"
     editor="$3"
+    edit_immutable="no"
+
+    readonly_filesystem() {
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: why is this needed twice? even if we need it to be redefined inside the script, we already exit in this case prior to calling the script

File: files/system/usr/bin/run0edit (hunk 3 above)
Comment (HastD) [reply files/system/usr/bin/run0edit:58]: Yes, a `--` argument can be used to treat the remaining arguments as positional, even if they start with a `-`. In this case since only one positional argument is accepted, it's just a way to handle the edge case of paths that start with a `-`. Probably not strictly necessary but it seems to be a pretty standard convention for this sort of command-line utility.

Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit:58]: makes sense

File: files/system/usr/bin/run0edit (hunk 4 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: This handles the case where the file is on a read-only filesystem, but neither the file nor its directory are readable at all by the user—in that case we can't detect that it's read-only until we're root.

Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit]: Ah I see

File: files/system/usr/bin/run0edit [hunk 5]
```diff
@@ -65,33 +109,89 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
+        fi
+    }
+
+    ask_immutable() {
+        if command -v lsattr >/dev/null; then
+            immutable_flag=$(lsattr -d -- "$1" 2>/dev/null | cut -d " " -f1 | tr -d "-" || echo "")
+        else
+            immutable_flag=""
+        fi
+        if [ "$immutable_flag" = i ]; then
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: should `i` be in quotes?

File: files/system/usr/bin/run0edit [hunk 6]
```diff
@@ -65,33 +109,89 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
+        else
+            false
+        fi
+    }
+
+    ask_immutable() {
+        if command -v lsattr >/dev/null; then
+            immutable_flag=$(lsattr -d -- "$1" 2>/dev/null | cut -d " " -f1 | tr -d "-" || echo "")
+        else
+            immutable_flag=""
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: is this required?

File: files/system/usr/bin/run0edit [hunk 7]
```diff
@@ -102,37 +202,73 @@ script='\
...
-        elif ! cp "$filename" "$tmpfile"; then
+            if readonly_filesystem "$filename"; then
+                echo "run0edit: $filename is on a read-only filesystem."
+                exit 1
+            fi
+            if ask_immutable "$filename"; then
+                edit_immutable="yes"
+            fi
+            if [ "$edit_immutable" = "no" ]; then
+                echo "run0edit: $filename is read-only."
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: does `edit_immutable` necessarily mean this? seems like it can also mean that the user opted to decline editing it

File: files/system/usr/bin/run0edit [hunk 8]
```diff
@@ -102,37 +202,73 @@ script='\
...
+            set -e
+            if [ "$edit_immutable" = "yes" ]; then
+                chattr +i -- "$directory"
+                echo "File created and immutable flag reapplied to directory."
+            fi
+            if [ "$cp_exit_status" != 0 ]; then
                 echo "run0edit: unable to write temporary file at $tmpfile to $filename"
                 exit 1
+            elif [ "$edit_immutable" = "yes" ] && ! cmp -s -- "$tmpfile" "$filename"; then
+                echo "WARNING: contents of $filename does not match contents of tempfile."
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: why is this needed?

File: files/system/usr/bin/run0edit (hunk 5 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: Doesn't make a difference, it doesn't have any special characters so it'll be treated as a string literal either way...

File: files/system/usr/bin/run0edit (hunk 6 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: Oh, good point, if `lsattr` isn't found, it'll just fall back to the `echo ""` path anyway... I'll remove the unnecessary conditional.

File: files/system/usr/bin/run0edit [hunk 9]
```diff
@@ -102,37 +202,73 @@ script='\
...
         if [ -s "$tmpfile" ]; then
-            touch "$filename"
-            if ! cp "$tmpfile" "$filename"; then
+            [ "$edit_immutable" = "yes" ] && chattr -i -- "$directory"
+            touch -- "$filename"
+            set +e
+            cp -- "$tmpfile" "$filename"
+            cp_exit_status="$?"
+            set -e
+            if [ "$edit_immutable" = "yes" ]; then
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: this works but is error-prone. It would be better to explicitly tie the change in attrs together in the flow. for example if we had the logic for copying the tempfile in a function like `copy_to_dest`

then we could do something like:

```
if edit_immutable
chattr -i
copy_to_dest
chattr +i
else
copy_to_dest
fi
```

this is a picky nit but for something so critical I'm trying to be as picky as possible

File: files/system/usr/bin/run0edit [hunk 10]
```diff
@@ -65,33 +109,89 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
-            exit 1
+            if readonly_filesystem "$directory"; then
+                echo "run0edit: $directory is on a read-only filesystem."
+                exit 1
+            fi
+            if ask_immutable "$directory"; then
+                edit_immutable="yes"
+            fi
+            if [ "$edit_immutable" = "no" ]; then
+                echo "run0edit: $directory is read-only."
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: same thing here: does edit_immutable necessarily mean this? seems like it can also mean that the user opted to decline editing it

File: files/system/usr/bin/run0edit (hunk 7 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: If the user declined removing the immutable flag, then the file is read-only, so this message is correct.

File: files/system/usr/bin/run0edit (hunk 8 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: There's a brief window between when the file is overwritten with its new contents and when the immutable flag is reapplied during which another unprivileged program could edit the file. I can't think of any way to get rid of this race condition entirely, so as a best-effort mitigation, we just compare it to the temporary file contents *after* reapplying the immutable flag to make sure they match.

File: files/system/usr/bin/run0edit (hunk 9 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: Good idea, I'll do that (and likewise for the "create new file" case).

File: files/system/usr/bin/run0edit (hunk 7 above)
Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit]: Doh, right

File: files/system/usr/bin/run0edit (hunk 8 above)
Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit]: makes sense

Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit]: @HastD could we use `flock`?

**Comment by HastD** [comment]
Just FYI, I refactored the script so that, in the [run0edit repo](https://github.com/HastD/run0edit), the main script and inner (privileged) script are now in separate files and are merged into the final `run0edit` script by a (very simple) Python script, `build.py`.

So, you may find it easier to review the inner script at https://github.com/HastD/run0edit/blob/main/run0edit-inner.sh since it has proper syntax highlighting—the `run0edit` file in this PR is produced simply by substituting the contents of `run0edit-inner.sh` into `run0edit-main.sh` in the appropriate place and applying the patch at `run0edit-secureblue.patch` (which doesn't make any changes to the inner script anyway).

File: files/system/usr/bin/run0edit (hunk 10 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: I did end up changing the message here as part of the control flow improvements based on your suggestions.

File: files/system/usr/bin/run0edit (hunk 8 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: Isn't the lock set by `flock` optional, in the sense that programs can simply ignore it if they choose? So I don't think it'd help here.

File: files/system/usr/bin/run0edit [hunk 11]
```diff
@@ -65,93 +109,204 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
+    ask_immutable_response=''
+    read -r ask_immutable_response
+    case "$ask_immutable_response" in
+        y*|Y*) true ;;
+        *) false ;;
+    esac
+}
+
+# If the file does not exist, ensure the directory exists and is writable.
+if [ ! -e "$filename" ]; then
```
Comment (ShadowSlayer1441) [inline files/system/usr/bin/run0edit]: Why do we only check the directory of the file if it doesn't exist? Can you edit files within an immutable directory? I thought you couldn't. (I can't check rn.)

Edit: nvm, apparently you can edit files within immutable directories.

File: files/system/usr/bin/run0edit (hunk 8 above)
Comment (ShadowSlayer1441) [reply files/system/usr/bin/run0edit]: I said it was optional earlier on discord, but reading the flock man page again, maybe it is enforced:

> Since Linux 5.5, flock() locks are emulated with SMB byte-range
       locks on the entire file.  Similarly to NFS, this means that
       fcntl(2) and flock() locks interact with one another.  Another
       important side-effect is that the **locks are not advisory** anymore:
       any IO on a locked file will always fail with EACCES when done
       from a separate file descriptor.  This difference originates from
       the design of locks in the SMB protocol, which provides mandatory
       locking semantics. 

Never mind, I tested this, it doesn't seem to be true. While this was running I was able to open and save the a edit with no problem. I checked with lsof that this program actually locked the file, but it didn't effect anything I tried.
```
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdio.h>
#include <sys/file.h>

int main() {
	int fd = open("/var/home/user/flock_test_file", O_RDWR);
	if (fd == -1) {
		printf("fd failed\n");
		return 1;
	}

	if (flock(fd, LOCK_EX) == -1) {
		printf("flock failed\n");
		return 1;
	}

	pause();

	flock(fd, LOCK_UN);
	close(fd);
	return 0;
}
```

File: files/system/usr/bin/run0edit [hunk 12]
```diff
@@ -65,93 +109,204 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
-        directory="$(dirname "$filename")"
-        if [ ! -d "$directory" ]; then
-            echo "run0edit: invalid argument: directory does not exist"
+script=$(cat <<-'END_OF_INNER_SCRIPT'
+#!/bin/sh
+# shellcheck disable=SC3040
+( set -o pipefail 2> /dev/null ) && set -o pipefail
+set -eu
+# shellcheck disable=SC2016
+PATH=$(command -p env -i sh -c 'echo "$PATH"')
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit:138]: what's then purpose of this?

File: files/system/usr/bin/run0edit [hunk 13]
```diff
@@ -65,93 +109,204 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
+            touch -- "$filename"
+            set +e
+            cp -- "$tmpfile" "$filename"
+            cp_exit_status="$?"
+            set -e
+            chattr +i -- "$directory"
+            echo 'File created and immutable flag reapplied to directory.'
         else
-            echo "run0edit: $filename not created"
+            touch -- "$filename"
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: this chunk is duplicated and should be factored

File: files/system/usr/bin/run0edit (hunk 12 above)
Comment (HastD) [reply files/system/usr/bin/run0edit:138]: Same thing as in the main script, resets `PATH` to a hard-coded default value.

File: files/system/usr/bin/run0edit (hunk 13 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: I split out some chunks of code into separate functions, that should get rid of most of the code duplication and make the control flow easier to follow.

File: files/system/usr/bin/run0edit [hunk 14]
```diff
@@ -65,93 +109,181 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
 else
-    echo "Editor not found. Please install either nano or vi, or write the path to" >&2
-    echo "the text editor of your choice to /etc/run0edit/editor.conf" >&2
+    echo 'Editor not found. Please install either nano or vi, or write the path to' >&2
+    echo 'the text editor of your choice to /etc/run0edit/editor.conf' >&2
     exit 1
 fi
 
+readonly_filesystem() {
+    if [ -r "$1" ] && command -pv findmnt >/dev/null; then
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: this could probably be slimmed down to a one-liner, having a separate check for whether the findmnt command exists is redundant I believe

File: files/system/usr/bin/run0edit (hunk 12 above)
Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit:138]: why not just PATH=""?

File: files/system/usr/bin/run0edit [hunk 15]
```diff
@@ -65,93 +109,181 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
-    # If the file does not exist, ensure the directory exists and is writable.
-    if [ ! -e "$filename" ]; then
-        directory="$(dirname "$filename")"
-        if [ ! -d "$directory" ]; then
-            echo "run0edit: invalid argument: directory does not exist"
-            exit 1
-        elif [ ! -w "$directory" ]; then
-            echo "run0edit: $directory is read-only."
-            exit 1
+readonly_filesystem() {
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit:143]: having these duplicated still smells off to me, but I'm not sure of a better way. passing it between the outer and inner scripts would be a vulnerability...

File: files/system/usr/bin/run0edit [hunk 16]
```diff
@@ -65,93 +109,181 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
+        exit 1
+    fi
+}
+
+copy_to_dest() {
+    readonly chattr_target_path="$1"
+    readonly edit_immutable="$2"
+    if [ "$edit_immutable" = 'yes' ]; then
+        chattr -i -- "$chattr_target_path"
+        touch -- "$filename"
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit]: im not sure where the previous comment went but there's still code duplication here. we can move this to:

```bash
copy_to_dest() {
       touch -- "$filename"
        set +e
        cp -- "$tmpfile" "$filename"
        cp_exit_status="$?"
        set -e
}
```

and rename copy_to_dest to copy_immutable_handler, with:

```bash
copy_immutable_handler() {
    readonly chattr_target_path="$1"
    readonly edit_immutable="$2"
    if [ "$edit_immutable" = 'yes' ]; then
        chattr -i -- "$chattr_target_path"
        copy_to_dest
        chattr +i -- "$chattr_target_path"
        echo 'Immutable flag reapplied.'
    else
        copy_to_dest
    fi
...
}
```

Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit]: the exit status will have to be passed up

File: files/system/usr/bin/run0edit [hunk 17]
```diff
@@ -65,93 +109,181 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
...
+temp_filename=$(mktemp --tmpdir -- "${trunc_filename}.XXXXXXXXXX")
 
 # Inner script to be run with elevated privileges:
 # shellcheck disable=SC2016
-script='\
-    filename="$1"
-    tmpfile="$2"
-    editor="$3"
+script=$(cat <<-'END_OF_INNER_SCRIPT'
+#!/bin/sh
```
Comment (RoyalOughtness) [inline files/system/usr/bin/run0edit:133]: does this imply that we intend to support any shell? do we want this?

File: files/system/usr/bin/run0edit (hunk 12 above)
Comment (HastD) [reply files/system/usr/bin/run0edit:138]: Because then `/usr/bin` wouldn't be in `PATH` and the shell wouldn't be able to find standard utilities?

File: files/system/usr/bin/run0edit (hunk 17 above)
Comment (HastD) [reply files/system/usr/bin/run0edit:133]: Yes, I've written the script to work in any POSIX-compatible shell; that's been the case since the first version. My intent is for it to be usable on any Linux system using systemd, so I've avoided using any bash-specific extensions.

File: files/system/usr/bin/run0edit (hunk 12 above)
Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit:138]: we could hardcode it to /usr/bin? do we need anything else

File: files/system/usr/bin/run0edit (hunk 14 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: Good point, and likewise the `[ -r "$1" ]` check can be dropped as `findmnt` will simply fail if the file isn't readable.

File: files/system/usr/bin/run0edit (hunk 15 above)
Comment (HastD) [reply files/system/usr/bin/run0edit:143]: I can't think of a better way either, and I don't think it's a big deal to duplicate a single one-liner function (which is what it'll be now).

File: files/system/usr/bin/run0edit (hunk 16 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: It's just a few lines duplicated here and I think it's clearer this way vs splitting off a second function and passing a value between them, especially given how clunky shell functions are (they don't really have return values). Actually, I think it'd be preferable to just duplicate the `edit_immutable` check, like this:
```sh
    if [ "$edit_immutable" = 'yes' ]; then
        chattr -i -- "$chattr_target_path"
    fi
    touch -- "$filename"
    set +e
    cp -- "$tmpfile" "$filename"
    cp_exit_status="$?"
    set -e
    if [ "$edit_immutable" = 'yes' ]; then
        chattr +i -- "$chattr_target_path"
        echo 'Immutable flag reapplied.'
    fi
```

Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit]: that's what you had before :smile: but yes I agree that's preferable if we're not using a separate function

my concern still stands though about later down the line someone could inadvertently change the control flow by modifying edit_immutable in-between, but it's a minor concern

**Review by RoyalOughtness — APPROVED** [review]

File: files/system/usr/bin/run0edit (hunk 16 above)
Comment (HastD) [reply files/system/usr/bin/run0edit]: `edit_immutable` is declared as `readonly` though so it'd be an error to try to modify it after initial assignment.

Comment (RoyalOughtness) [reply files/system/usr/bin/run0edit]: Oh nvm, my concern is mitigated already then

**Review by ShadowSlayer1441 — APPROVED** [review]
